### DISCUSSION TITLE
Reset device

### DISCUSSION THREAD

**Discussion by xDarkDelux** [mikke89/RmlUi#410] [Q&A]
i'm trying to use rmlui with directx9 renderer and i can't figure out how to properly recreate the device in rmlui.. i.e. i need to implement 2 functions specifically for rmlui - DeviceLost and DeviceFound, please help me understand which functions i need to call in RmlUI to delete everything links to old objects created by the old device, ReleaseTextures alone will obviously not do here, I would not want to completely reload the entire rmlui or all windows

**Comment by mikke89** [comment]
In addition to the `Rml::ReleaseTextures` as you mentioned, there is also the `Rml::ReleaseCompiledGeometry`. I believe those should cover all resources.

**Reply by xDarkDelux** [reply]
unfortunately this did not help, here is what I do immediately after recreating the device:

```
void RmlUI::DeviceFound()
{
	Rml::ReleaseTextures(&render_interface);
	Rml::ReleaseCompiledGeometry();

	for (INT i = 0; i < GInterfaceData.wnd_arr.Num(); ++i)
	{
		if (Window* wnd = GInterfaceData.wnd_arr(i))
		{
			wnd->ReloadStyle();
		}
	}

	is_valid_device = TRUE;
}
```

And I see that everything seems to be ok, the textures are reloaded and everything works. input) - I get such a crit error, and this error pops up in different situations in the same way, I can’t understand what the problem is:
![Screenshot_21](https://user-images.githubusercontent.com/30740744/217810288-46c6f048-095b-4dc1-ab32-d6347c7b8519.png)
![Screenshot_22](https://user-images.githubusercontent.com/30740744/217810417-ef527f55-046a-4756-8a24-6839644fea6a.png)

**Reply by xDarkDelux** [reply]
this error after recreating the device can be in different situations and with calls to compile geometry from other places, but the error itself is always like this

**Reply by mikke89** [reply]
Hmm, I don't know what's going on, can't really help you here, it certainly looks like some issue on the client side.

**Comment by xDarkDelux** [comment]
what do you mean by "client side issue"? at least roughly can you advise what to watch and what to check?

**Reply by mikke89** [reply]
You really just need to break out your debugger, and look through your code. Figuring out why `d3d9_vertices` is null would be a good place to start, and backtrack from there.
